Python/Product/PythonTools/PythonTools/Environments/EnvironmentSwitcherManager.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Internal.VisualStudio.Shell;
+using Microsoft.PythonTools.Analysis;
+using Microsoft.PythonTools.Infrastructure;
+using Microsoft.PythonTools.Interpreter;
+using Microsoft.PythonTools.Project;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell.Interop;
+
+// TODO: handle the workspace scenario
+
+namespace Microsoft.PythonTools.Environments {
+    /// <summary>
+    /// This keeps track of selection and other changes in the IDE and caches
+    /// the data to be displayed in the environment status bar switcher,
+    /// such as the list of all environments, the current environment,
+    /// and the current project.
+    /// </summary>
+    sealed class EnvironmentSwitcherManager : IVsSelectionEvents, IDisposable {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IInterpreterOptionsService _optionsService;
+        private readonly IInterpreterRegistryService _registryService;
+        private readonly IVsMonitorSelection _monitorSelection;
+        private uint _selectionEventsCookie;
+        private IVsHierarchy _previousHier;
+        private PythonProjectNode _project;
+
+        public event EventHandler EnvironmentsChanged;
+
+        public EnvironmentSwitcherManager(IServiceProvider serviceProvider) {
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            _optionsService = serviceProvider.GetComponentModel().GetService<IInterpreterOptionsService>();
+            _registryService = serviceProvider.GetComponentModel().GetService<IInterpreterRegistryService>();
+            _monitorSelection = serviceProvider.GetService(typeof(IVsMonitorSelection)) as IVsMonitorSelection;
```

Comment:
Probably `SVsMonitorSelection`

Reply:
I thought that existed too, but I guess it doesn't. It's not in `Microsoft.VisualStudio.Shell.Interop`.